### PR TITLE
ERMS-2338, ERMS-2353, ERMS-2356

### PR DESCRIPTION
**Pull request by Moetez16**

### CODE REVIEW

File: app/grails-app/controllers/com/k_int/kbplus/MyInstitutionController.groovy [hunk 1]
```diff
@@ -2626,6 +2626,8 @@ AND EXISTS (
         result.surveyConfig = SurveyConfig.get(params.id)
         result.surveyInfo = result.surveyConfig.surveyInfo
 
+        result.surveyResults = SurveyResult.findAllByParticipantAndSurveyConfig(result.institution, result.surveyConfig).sort { it?.surveyConfig?.configOrder }
```
Comment (croqueGrec09) [inline app/grails-app/controllers/com/k_int/kbplus/MyInstitutionController.groovy:2629]: to be or not to be => ?

File: app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy [hunk 2]
```diff
@@ -2854,9 +2854,10 @@ class SubscriptionController extends AbstractDebugController {
             if(params.singleTitle) {
                 IssueEntitlement ie = IssueEntitlement.get(params.singleTitle)
                 def tipp = ie.tipp
+
                 try {
 
-                    if(subscriptionService.addEntitlement(result.subscriptionInstance, tipp.gokbId, ie, true, ie_accept_status)) {
+                    if(subscriptionService.addEntitlement(result.subscriptionInstance, tipp.gokbId, ie, ie.priceItem, ie_accept_status)) {
```
Comment (croqueGrec09) [inline app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy:2860]: where does ie.priceItem comes from? This is actually a flag in addEntitlements, indicating whether price information should be considered or not!

File: app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy [hunk 3]
```diff
@@ -3001,9 +3002,10 @@ class SubscriptionController extends AbstractDebugController {
                 IssueEntitlement ie = IssueEntitlement.findById(ieID)
                 def tipp = ie.tipp
 
+
                 if(tipp) {
                     try {
-                        if (subscriptionService.addEntitlement(result.subscriptionInstance, tipp.gokbId, ie, true, ie_accept_status)) {
+                        if (subscriptionService.addEntitlement(result.subscriptionInstance, tipp.gokbId, ie, ie.priceItem, ie_accept_status)) {
```
Comment (croqueGrec09) [inline app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy:3008]: same as above

File: app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy [hunk 4]
```diff
@@ -1847,7 +1854,7 @@ class SurveyController {
 
             result.surveyConfigs.each { config ->
 
-                if(!config?.pickAndChoose) {
+
                     config.orgs?.org?.each { org ->
 
                             config?.surveyProperties?.each { property ->
```
Comment (croqueGrec09) [inline app/grails-app/controllers/com/k_int/kbplus/SurveyController.groovy:1860]: How can config be null at this place? In line 1858, we presume already that config is *not* null!

File: app/grails-app/services/de/laser/SubscriptionService.groovy [hunk 5]
```diff
@@ -172,6 +172,34 @@ class SubscriptionService {
         validSubChilds
     }
 
+    List getValidSurveySubChilds(Subscription subscription) {
+        def validSubChilds = Subscription.findAllByInstanceOfAndStatusInList(
+                subscription,
+                [SUBSCRIPTION_CURRENT, SUBSCRIPTION_UNDER_PROCESS_OF_SELECTION]
+        )
+        validSubChilds = validSubChilds?.sort { a, b ->
```
Comment (croqueGrec09) [inline app/grails-app/services/de/laser/SubscriptionService.groovy:180]: make sort in if()

File: app/grails-app/services/de/laser/SubscriptionService.groovy [hunk 6]
```diff
@@ -172,6 +172,34 @@ class SubscriptionService {
...
+            def sa = a.getSubscriber()
+            def sb = b.getSubscriber()
+            (sa?.sortname ?: sa?.name ?: "")?.compareTo((sb?.sortname ?: sb?.name ?: ""))
```
Comment (croqueGrec09) [inline app/grails-app/services/de/laser/SubscriptionService.groovy:183]: then the superfluous ? can disappear (sa and sb can then not be null any more)

File: app/grails-app/taglib/de/laser/SemanticUiTagLib.groovy [hunk 7]
```diff
@@ -1083,7 +1083,16 @@ class SemanticUiTagLib {
 
         if (surveyConfig?.pickAndChoose) {
             def finishDate = SurveyOrg.findBySurveyConfigAndOrg(surveyConfig, participant).finishDate
-            if (finishDate) {
+            List surveyResults = SurveyResult.findAllByParticipantAndSurveyConfig(participant, surveyConfig)
+
+            boolean finish = false
+
+            if (surveyResults) {
+                finish = finishDate && surveyResults?.finishDate?.contains(null)
```
Comment (croqueGrec09) [inline app/grails-app/taglib/de/laser/SemanticUiTagLib.groovy:1091]: to be or not to be => ?

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 8]
```diff
@@ -40,11 +40,13 @@
                it?.participant.sortname
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:40]: can it be really null at this place?

File: app/grails-app/views/templates/filter/_orgFilterTable.gsp [hunk 9]
```diff
@@ -558,6 +558,11 @@
 
                         <g:if test="${orgSub?.isCurrentMultiYearSubscriptionNew()}">
                             <g:message code="surveyOrg.perennialTerm.available"/>
+                            <br>
+                            <g:link controller="subscription" action="show"
+                                    id="${orgSub?.id}">
```
Comment (klodav) [inline app/grails-app/views/templates/filter/_orgFilterTable.gsp:563]: ? <- orgSub is not null here

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 10]
```diff
@@ -139,64 +141,64 @@
                         <g:set var="surveyOrg"
                                value="${SurveyOrg.findBySurveyConfigAndOrg(resultProperty?.surveyConfig, participant)}"/>
 
-                        <g:if test="${!surveyOrg?.existsMultiYearTerm()}">
+                        <g:if test="${surveyOrg?.existsMultiYearTerm() && resultProperty?.surveyConfig.subSurveyUseForTransfer}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:144]: can resultProperty be null at this place?

File: app/grails-app/views/templates/filter/_orgFilterTable.gsp [hunk 11]
```diff
@@ -574,6 +579,14 @@
                     <g:if test="${existSubforOrg}">
                         <g:if test="${orgSub?.isCurrentMultiYearSubscriptionNew()}">
                             <g:message code="surveyOrg.perennialTerm.available"/>
+                            <br>
+                            <g:link controller="subscription" action="show"
+                                    id="${orgSub?.id}">
```
Comment (klodav) [inline app/grails-app/views/templates/filter/_orgFilterTable.gsp:584]: ? <- orgSub is not null here

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 12]
```diff
@@ -139,64 +141,64 @@
...
-                                                 class="la-overflow la-ellipsis"/>
-                                <g:if test="${resultProperty?.urlValue}">
-                                    <semui:linkIcon/>
-                                </g:if>
-                            </g:elseif>
-                            <g:elseif test="${resultProperty?.type?.type == RefdataValue.toString()}">
-                                <semui:xEditableRefData owner="${resultProperty}" type="text" field="refValue"
-                                                        config="${resultProperty.type?.refdataCategory}"/>
-                            </g:elseif>
                             <g:if test="${resultProperty?.comment}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:148]: same as above

File: app/grails-app/views/templates/filter/_orgFilterTable.gsp [hunk 13]
```diff
@@ -592,6 +605,11 @@
...
```
Comment (klodav) [inline app/grails-app/views/templates/filter/_orgFilterTable.gsp:610]: ? <- orgSub is not null here

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 14]
```diff
@@ -139,64 +141,64 @@
...
                                 <span class="la-long-tooltip la-popup-tooltip la-delay" data-position="right center"
                                       data-content="${resultProperty?.comment}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:150]: same as above

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 15]
```diff
@@ -139,64 +141,64 @@
...
+                            <g:else>
 
-                            <g:if test="${resultProperty?.type?.id == RDStore.SURVEY_PROPERTY_PARTICIPATION.id && resultProperty?.getResult() == RDStore.YN_NO.getI10n('value')}">
-                                <span class="la-long-tooltip la-popup-tooltip la-delay" data-position="top right"
-                                      data-variation="tiny"
-                                      data-content="${message(code: 'surveyResult.particiption.terminated')}">
-                                    <i class="minus circle big red icon"></i>
-                                </span>
-                            </g:if>
+                                <g:if test="${resultProperty?.type?.name == "Participation"}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:157]: same as above

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 16]
```diff
@@ -139,64 +141,64 @@
...
+                                    <g:set var="resultPropertyParticipation" value="${resultProperty}"/>
+                                </g:if>
 
-                        </g:if>
-                        <g:else>
+                                <g:if test="${resultProperty?.type?.type == Integer.toString()}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:161]: same as above

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 17]
```diff
@@ -139,64 +141,64 @@
...
+                                </g:elseif>
+                                <g:if test="${resultProperty?.comment}">
+                                    <span class="la-long-tooltip la-popup-tooltip la-delay" data-position="right center"
+                                          data-content="${resultProperty?.comment}">
+                                        <i class="question circle icon"></i>
+                                    </span>
+                                </g:if>
 
-                            <g:message code="surveyOrg.perennialTerm.available"/>
+                                <g:if test="${resultProperty?.type?.id == RDStore.SURVEY_PROPERTY_PARTICIPATION.id && resultProperty?.getResult() == RDStore.YN_NO.getI10n('value')}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:192]: same as above (and many many other places, too ...)

File: app/grails-app/views/survey/_evaluationSubscription.gsp [hunk 18]
```diff
@@ -277,7 +281,18 @@
                         <g:set var="surveyOrg"
                                value="${SurveyOrg.findBySurveyConfigAndOrg(resultProperty?.surveyConfig, participant)}"/>
 
-                        <g:if test="${!surveyOrg?.existsMultiYearTerm()}">
+                        <g:if test="${surveyOrg?.existsMultiYearTerm() && resultProperty?.surveyConfig.subSurveyUseForTransfer}">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/_evaluationSubscription.gsp:284]: null check of surveyOrg, resultProperty should not be null at this place (see above .........)

File: app/grails-app/views/survey/currentSurveysConsortia.gsp [hunk 19]
```diff
@@ -285,9 +285,12 @@
                         <g:if test="${surveyConfig}">
                             <g:link controller="survey" action="surveyParticipants" id="${surveyInfo.id}"
                                     params="[surveyConfigID: surveyConfig.id]" class="ui icon">
-                                <div class="ui circular ${participantsFinish?.size() == participantsTotal?.size() ? "green" : surveyConfig?.configFinish ? "yellow" : ""} label">
+                                    <div class="ui circular ${surveyConfig?.configFinish ? "yellow" : ""} label">
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/currentSurveysConsortia.gsp:288]: to be or not to be => ?

File: app/grails-app/views/templates/filter/_orgFilterTable.gsp [hunk 20]
```diff
@@ -558,6 +558,11 @@
 
                         <g:if test="${orgSub?.isCurrentMultiYearSubscriptionNew()}">
```
Comment (croqueGrec09) [inline app/grails-app/views/templates/filter/_orgFilterTable.gsp:559]: see above - should orgSub not be already not null?

File: app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy (hunk 2 above)
Comment (Moetez16) [reply app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy:2860]: It's come from ie. It is only for Survey so!

File: app/grails-app/views/templates/filter/_orgFilterTable.gsp (hunk 9 above)
Comment (Moetez16) [reply app/grails-app/views/templates/filter/_orgFilterTable.gsp:563]: Of course. Why not?

Comment (klodav) [reply app/grails-app/views/templates/filter/_orgFilterTable.gsp:563]: Line 559: **<g:if test="${orgSub?**.isCurrentMultiYearSubscriptionNew()}">

File: app/grails-app/views/survey/surveyTitlesEvaluation.gsp [hunk 21]
```diff
@@ -70,7 +89,7 @@
                     <h4><g:message code="surveyParticipants.hasAccess"/></h4>
 
                     <g:set var="surveyParticipantsHasAccess"
-                           value="${participantsFinish?.findAll { it?.hasAccessOrg() }?.sort {
+                           value="${participantsFinish.findAll { it?.hasAccessOrg() }?.sort {
```
Comment (croqueGrec09) [inline app/grails-app/views/survey/surveyTitlesEvaluation.gsp:92]: it?

**Pull request merged**